turtle.js

```diff
@@ -68,108 +71,102 @@ let turtle = _defaultTurtle();
  * draw the turtle and the current image if `redraw` is `true`.
  * for complicated drawings it is much faster to turn `redraw` off.
 */
-function drawIf() {
-    if (turtle.redraw) draw();
-}
+function drawIf() { if (turtle.redraw) draw() }
 
 /**
  * use canvas centered coordinates facing upwards
- * @param {CanvasRenderingContext2D} context
+ * @param {CanvasRenderingContext2D} ctx
  */
-const _centerCoords =  function(context) {
-    context.translate(context.canvas.width / 2, context.canvas.height / 2);
-    context.transform(1, 0, 0, -1, 0, 0);
+const _centerCoords = ctx => {
+    ctx.translate(ctx.canvas.width / 2, ctx.canvas.height / 2);
+    ctx.transform(1, 0, 0, -1, 0, 0);
 };
 
 /** draw the turtle and the current image */
 function draw() {
-    clearContext(turtleContext);
+    _clearCtx(_turtleCtx);
     if (turtle.visible) {
         const
             x = turtle.pos.x,
             y = turtle.pos.y;
 
-        turtleContext.save();
-        _centerCoords(turtleContext);
+        _turtleCtx.save();
+        _centerCoords(_turtleCtx);
         // move the origin to the turtle center
-        turtleContext.translate(x, y);
+        _turtleCtx.translate(x, y);
         // rotate about the center of the turtle
-        turtleContext.rotate(-turtle.angle);
+        _turtleCtx.rotate(-turtle.angle);
         // move the turtle back to its position
-        turtleContext.translate(-x, -y);
+        _turtleCtx.translate(-x, -y);
         // draw the turtle icon
         const icon = shapes.hasOwnProperty(turtle.shape) ?
             turtle.shape : "triangle";
-        turtleContext.beginPath();
+        _turtleCtx.beginPath();
         for (let i=0; i < shapes[icon].length; i++) {
             const coord = shapes[icon][i];
-            turtleContext[i==0 ? 'moveTo' : 'lineTo'](x+coord[0], y+coord[1]);
+            _turtleCtx[i==0 ? 'moveTo' : 'lineTo'](x+coord[0], y+coord[1]);
```

Comment:
This can be re-written like this:
`_turtleCtx[ i && 'lineTo' || 'moveTo' ](x + coord[0], y + coord[1]);`

Reply:
That's also good for minification (4B instead of 5B), but it may be confusing for some people. And it relies on the fact that both strings are truthy, instead of only having a single dependency on `i`

Reply:
> And it relies on the fact that both strings are truthy,

Actually for the pattern `cond && trueExp || falseExp` only the `trueExp` has to be truthy for the whole thing to work; otherwise we need to use the ternary operator `cond? trueExp : falseExp` instead.

Given that a non-empty string is truthy, `cond && trueExp || falseExp` pattern can be used as an alternative for the ternary operator in this particular case.

Regardless, having a special condition just for when iterator _i_ is `0` slows down that loop.

Given minification apparently isn't a consideration for this library, moving the `index = 0` case outta the loop would make things more performant:
```js
const
  icons = shapes[shapes.hasOwnProperty(turtle.shape)? turtle.shape:'triangle'],
  iconsLen = icons.length;

_turtleCtx.beginPath();
iconsLen && _turtleCtx.moveTo(x + icons[0][0], y + icons[0][1]);

for (let i = 0; ++i < iconsLen; ) {
  const [px, py] = icons[i];
  _turtleCtx.lineTo(x + px, y + py);
}

_turtleCtx.closePath();
```
I've also taken the opportunity to cache the _shapes{}[][]_ object 2d array as much as I could, so the loop will now need much less reference indirections to access all the coord values.

Reply:
That's so much better! I'm gonna implement it right now